### PR TITLE
Add static netcdf-c support to ESMF

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR adds support for ESMF to compile against static netcdf-c. This change supports NOAA/NWS R&D and operational needs.

### CODE REVIEW

**Review by jedwards4b — DISMISSED** [review]

**Review by climbfuji — DISMISSED** [review]

**Comment by skosukhin** [comment]
I repeat here  https://github.com/spack/spack/pull/34577#issuecomment-1355603797: please, don't merge this until my review. Next week, hopefully. At least the netcdf part is rather debatable: I don't see any reason yet to call pkg-config from the Python code.

**Review by alalazo — CHANGES_REQUESTED** [review]
Pending @skosukhin review

**Comment by skosukhin** [comment]
This is waiting for https://github.com/spack/spack/pull/34577.

I just wanted to mention #30371 here, which is partially reverted by this PR.

**Comment by alalazo** [comment]
> This is waiting for https://github.com/spack/spack/pull/34577.

Converting this to a draft to reflect the comment above

**Comment by AlexanderRichert-NOAA** [comment]
@skosukhin Can you please take a look again now that https://github.com/spack/spack/pull/34577 is merged? Based on some testing with the final version of that PR, it looks like these changes are still necessary (though of course I'm very open to the possibility that there's a better way to do it). Thanks.

**Comment by skosukhin** [comment]
@AlexanderRichert-NOAA could you, please, check whether #35466 works for you?

**Comment by AlexanderRichert-NOAA** [comment]
@skosukhin Yep, this looks great, thanks. I still need the modification to ESMF, so I'll remove the netcdf-fortran changes and then tweak the ESMF package.py to use "which" for nc-config rather than subprocess.

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (skosukhin) [inline var/spack/repos/builtin/packages/esmf/package.py]: Nitpicking: you can implement [`setup_build_environment`](https://spack-tutorial.readthedocs.io/en/latest/tutorial_advanced_packaging.html?highlight=setup_build_environment#set-environment-variables-in-your-own-package) to set the environment variable.

Comment (jedwards4b) [reply var/spack/repos/builtin/packages/esmf/package.py]: Maybe I'm missing something - but this change is already in a PR. https://github.com/spack/spack/pull/35249

File: var/spack/repos/builtin/packages/esmf/package.py
Comment (skosukhin) [inline var/spack/repos/builtin/packages/esmf/package.py]: So, the makefile knows about `nc-config` but you still have to call it below?

Comment (AlexanderRichert-NOAA) [reply var/spack/repos/builtin/packages/esmf/package.py]: Correct. ESMF's build system filters out much of the necessary output of `nc-config --libs`, so it doesn't handle static dependencies. I've discussed this with the developers and I believe it will be fixed in the future, but for now we either need to call `nc-config` ourselves or patch ESMF (and I think this is cleaner).

**Comment by AlexanderRichert-NOAA** [comment]
@alalazo If this looks okay can we get this merged? I've narrowed the scope of the PR thanks to @skosukhin but as I mentioned in a previous comment, ESMF still has a build system quirk (which will be fixed in future versions) that requires this addition.

**Comment by alalazo** [comment]
Fine with me to merge as soon as @skosukhin approves too

**Review by alalazo — DISMISSED** [review]

**Comment by AlexanderRichert-NOAA** [comment]
I believe @skosukhin's concern was about netcdf-fortran which was previously part of this PR but was removed. As discussed in a resolved conversation, this change is still needed due to how the ESMF build system parses the `nc-config` output. @alalazo can we please merge this ASAP **after** #35980 is merged?

**Comment by AlexanderRichert-NOAA** [comment]
Now that https://github.com/spack/spack/pull/35980 is merged, @skosukhin please let me know if you have further concerns (I believe they were all addressed regarding the netcdf-fortran piece that's since been removed). @alalazo I'd like to get this merged ASAP please.

**Comment by skosukhin** [comment]
I am fine with the changes in this PR but doesn't it depend (in a way) on #35466?

**Comment by AlexanderRichert-NOAA** [comment]
I can go either way. This PR doesn't strictly depend on it. Fingers crossed on #35466 getting merged soon as well.

**Review by alalazo — APPROVED** [review]

**Pull request merged**
